app/server/lib/OIDCConfig.ts

```diff
@@ -167,33 +306,58 @@ export class OIDCConfig {
         profile,
       }));
 
-      delete mreq.session.oidc;
+      mreq.session.oidc = {
+        idToken: tokenSet.id_token, // keep idToken for logout
+      };
       res.redirect(targetUrl ?? '/');
     } catch (err) {
       log.error(`OIDC callback failed: ${err.stack}`);
-      // Delete the session data even if the login failed.
+      if (Object.prototype.hasOwnProperty.call(err, 'response')) {
+        log.error(`Response received: ${err.response?.body ?? err.response}`);
+      }
+
+      // Delete entirely the session data when the login failed.
       // This way, we prevent several login attempts.
       //
       // Also session deletion must be done before sending the response.
-      delete mreq.session.oidc;
-      res.status(500).send(`OIDC callback failed.`);
+      try {
+        const mreq = this._getRequestWithSession(req);
+        delete mreq.session.oidc;
+      } catch (e) {
+        // _getRequestWithSession may fail because there is no session,
+        // in such a case, ignore the exception, just log some error.
+        log.warn('Could not retrieve session to delete the oidc protections');
+      }
+
+      await this._sendAppPage(req, res, {
+        path: 'error.html',
+        status: 500,
+        config: {
+          errPage: 'signin-failed',
+          errMessage: err.userFriendlyMessage
+        },
+      });
     }
   }
 
   public async getLoginRedirectUrl(req: express.Request, targetUrl: URL): Promise<string> {
-    const { codeVerifier, state } = await this._generateAndStoreConnectionInfo(req, targetUrl.href);
-    const codeChallenge = generators.codeChallenge(codeVerifier);
+    const protections = this._protectionManager.generate();
+    const mreq = this._getRequestWithSession(req);
+
+    mreq.session.oidc = {
+      targetUrl: targetUrl.href,
+      ...protections
+    };
 
-    const authUrl = this._client.authorizationUrl({
+    return this._client.authorizationUrl({
       scope: process.env.GRIST_OIDC_IDP_SCOPES || 'openid email profile',
-      code_challenge: codeChallenge,
-      code_challenge_method: 'S256',
-      state,
+      acr_values: this._acrValues,
+      ...this._protectionManager.forgeProtectionParamsForAuthUrl(protections),
     });
-    return authUrl;
   }
 
   public async getLogoutRedirectUrl(req: express.Request, redirectUrl: URL): Promise<string> {
+    const mreq = this._getRequestWithSession(req);
```

Comment:
This can throw an exception, but that seems counterproductive. If session is missing, it's OK to just omit `id_token_hint`, no?